File: WorldFeed/Server/src/WorldFeed/Client/Clients/UploadClient.cs
namespace WorldFeed.Client.Clients
{
    using System.Threading.Tasks;

    using WorldFeed.Common.Client.Validators;
    using WorldFeed.Common.Public;
    using WorldFeed.Common.Public.Client.Clients;
    using WorldFeed.Common.Public.Client.Requesters;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Models.Interfaces.DTO;
    using WorldFeed.Common.Public.Parameters.Upload;

    public class UploadClient : IUploadClient
    {
        private readonly ITwitterClient client;
        private readonly IUploadRequester uploadRequester;

        public UploadClient(ITwitterClient client)
        {
            this.client = client;
            this.uploadRequester = client.Raw.Upload;
        }

        public IUploadClientParametersValidator ParametersValidator => this.client.ParametersValidator;

        public Task<IMedia> UploadBinaryAsync(byte[] binary)
        {
            return UploadBinaryAsync(new UploadBinaryParameters(binary));
        }

        public async Task<IMedia> UploadBinaryAsync(IUploadParameters parameters)
        {
            var chunkUploadResult = await this.uploadRequester.UploadBinaryAsync(parameters).ConfigureAwait(false);
            return chunkUploadResult.Media;
        }

        public Task<IMedia> UploadTweetImageAsync(byte[] binary)
        {
            return UploadTweetImageAsync(new UploadTweetImageParameters(binary));
        }

        public Task<IMedia> UploadTweetImageAsync(IUploadTweetImageParameters parameters)
        {
            return UploadBinaryAsync(parameters);
        }

        public Task<IMedia> UploadMessageImageAsync(byte[] binary)
        {
            return UploadMessageImageAsync(new UploadMessageImageParameters(binary));
        }

        public Task<IMedia> UploadMessageImageAsync(IUploadMessageImageParameters parameters)
        {
            return UploadBinaryAsync(parameters);
        }

        public Task<IMedia> UploadTweetVideoAsync(byte[] binary)
        {
            return UploadTweetVideoAsync(new UploadTweetVideoParameters(binary));
        }

        public Task<IMedia> UploadTweetVideoAsync(IUploadTweetVideoParameters parameters)
        {
            return UploadBinaryAsync(parameters);
        }

        public Task<IMedia> UploadMessageVideoAsync(byte[] binary)
        {
            return UploadMessageVideoAsync(new UploadMessageVideoParameters(binary));
        }

        public Task<IMedia> UploadMessageVideoAsync(IUploadMessageVideoParameters parameters)
        {
            return UploadBinaryAsync(parameters);
        }

        public Task AddMediaMetadataAsync(IMediaMetadata metadata)
        {
            return AddMediaMetadataAsync(new AddMediaMetadataParameters(metadata.MediaId)
            {
                AltText = metadata.AltText
            });
        }

        public async Task AddMediaMetadataAsync(IAddMediaMetadataParameters parameters)
        {
            await this.uploadRequester.AddMediaMetadataAsync(parameters).ConfigureAwait(false);
        }

        public async Task<IUploadedMediaInfo> GetVideoProcessingStatusAsync(IMedia media)
        {
            var twitterResult = await this.uploadRequester.GetVideoProcessingStatusAsync(media).ConfigureAwait(false);
            return twitterResult.Model;
        }

        public Task WaitForMediaProcessingToGetAllMetadataAsync(IMedia media)
        {
            return this.uploadRequester.WaitForMediaProcessingToGetAllMetadataAsync(media);
        }
    }
}

File: WorldFeed/Client/src/app/worldfeed/Client/Clients/UploadClient.ts
import {IUploadClient} from "../../../core/Public/Client/Clients/IUploadClient";
import {IUploadRequester} from "../../../core/Public/Client/Requesters/IUploadRequester";
import {IUploadClientParametersValidator} from "../../../core/Core/Client/Validators/UploadClientParametersValidator";
import {IMedia} from "../../../core/Public/Models/Interfaces/IMedia";
import {IUploadParameters, UploadBinaryParameters} from "../../../core/Public/Parameters/Upload/UploadBinaryParameters";
import {IUploadTweetImageParameters, UploadTweetImageParameters} from "../../../core/Public/Parameters/Upload/UploadTweetImageParameters";
import {IUploadMessageImageParameters, UploadMessageImageParameters} from "../../../core/Public/Parameters/Upload/UploadMessageImageParameters";
import {IUploadTweetVideoParameters, UploadTweetVideoParameters} from "../../../core/Public/Parameters/Upload/UploadTweetVideoParameters";
import {IUploadMessageVideoParameters, UploadMessageVideoParameters} from "../../../core/Public/Parameters/Upload/UploadMessageVideoParameters";
import {AddMediaMetadataParameters, IAddMediaMetadataParameters} from "../../../core/Public/Parameters/Upload/AddMediaMetadataParameters";
import {IUploadedMediaInfo} from "../../../core/Public/Models/Interfaces/DTO/IUploadedMediaInfo";
import {ITwitterClient} from "../../../core/Public/ITwitterClient";
import {IMediaMetadata} from "../../../core/Public/Models/Interfaces/DTO/IMediaMetadata";

export class UploadClient implements IUploadClient {
  private readonly _client: ITwitterClient;
  private readonly _uploadRequester: IUploadRequester;

  constructor(client: ITwitterClient) {
    this._client = client;
    this._uploadRequester = client.Raw.upload;
  }

  get parametersValidator(): IUploadClientParametersValidator {
    return this._client.ParametersValidator;
  }

  public async uploadBinaryAsync(binaryOrParameters: number[] | IUploadParameters): Promise<IMedia> {
    let parameters: IUploadParameters;
    if (this.isIUploadParameters(binaryOrParameters)) {
      parameters = binaryOrParameters;
    } else {
      parameters = new UploadBinaryParameters(binaryOrParameters);
    }

    let chunkUploadResult = await this._uploadRequester.uploadBinaryAsync(parameters); // .ConfigureAwait(false);
    return chunkUploadResult.media;
  }

  public uploadTweetImageAsync(binaryOrParameters: number[] | IUploadTweetImageParameters): Promise<IMedia> {
    let parameters: IUploadTweetImageParameters;
    if (this.isIUploadTweetImageParameters(binaryOrParameters)) {
      parameters = binaryOrParameters;
    } else {
      parameters = new UploadTweetImageParameters(binaryOrParameters);
    }

    return this.uploadBinaryAsync(parameters);
  }

  public uploadMessageImageAsync(binaryOrParameters: number[] | IUploadMessageImageParameters): Promise<IMedia> {
    let parameters: IUploadMessageImageParameters;
    if (this.isIUploadMessageImageParameters(binaryOrParameters)) {
      parameters = binaryOrParameters;
    } else {
      parameters = new UploadMessageImageParameters(binaryOrParameters);
    }

    return this.uploadBinaryAsync(parameters);
  }

  public uploadTweetVideoAsync(binaryOrParameters: number[] | IUploadTweetVideoParameters): Promise<IMedia> {
    let parameters: IUploadTweetVideoParameters;
    if (this.isIUploadTweetVideoParameters(binaryOrParameters)) {
      parameters = binaryOrParameters;
    } else {
      parameters = new UploadTweetVideoParameters(binaryOrParameters);
    }

    return this.uploadBinaryAsync(parameters);
  }

  public uploadMessageVideoAsync(binaryOrParameters: number[] | IUploadMessageVideoParameters): Promise<IMedia> {
    let parameters: IUploadMessageVideoParameters;
    if (this.isIUploadMessageVideoParameters(binaryOrParameters)) {
      parameters = binaryOrParameters;
    } else {
      parameters = new UploadMessageVideoParameters(binaryOrParameters);
    }

    return this.uploadBinaryAsync(parameters);
  }

  public async addMediaMetadataAsync(metadataOrParameters: IMediaMetadata | IAddMediaMetadataParameters): Promise<void> {
    let parameters: IAddMediaMetadataParameters;
    if (this.isIAddMediaMetadataParameters(metadataOrParameters)) {
      parameters = metadataOrParameters;
    } else {
      parameters = new AddMediaMetadataParameters(metadataOrParameters.mediaId);
      parameters.altText = metadataOrParameters.altText;
    }

    await this._uploadRequester.addMediaMetadataAsync(parameters); // .ConfigureAwait(false);
  }

  public async getVideoProcessingStatusAsync(media: IMedia): Promise<IUploadedMediaInfo> {
    let twitterResult = await this._uploadRequester.getVideoProcessingStatusAsync(media); // .ConfigureAwait(false);
    return twitterResult.model;
  }

  public waitForMediaProcessingToGetAllMetadataAsync(media: IMedia): Promise<void> {
    return this._uploadRequester.waitForMediaProcessingToGetAllMetadataAsync(media);
  }

  private isIUploadParameters(binaryOrParameters: number[] | IUploadParameters): binaryOrParameters is IUploadParameters {
    return (binaryOrParameters as IUploadParameters).binary !== undefined;
  }

  private isIUploadTweetImageParameters(binaryOrParameters: number[] | IUploadTweetImageParameters): binaryOrParameters is IUploadTweetImageParameters {
    return (binaryOrParameters as IUploadTweetImageParameters).binary !== undefined;
  }

  private isIUploadMessageImageParameters(binaryOrParameters: number[] | IUploadMessageImageParameters): binaryOrParameters is IUploadMessageImageParameters {
    return (binaryOrParameters as IUploadMessageImageParameters).binary !== undefined;
  }

  private isIUploadTweetVideoParameters(binaryOrParameters: number[] | IUploadTweetVideoParameters): binaryOrParameters is IUploadTweetVideoParameters {
    return (binaryOrParameters as IUploadTweetVideoParameters).binary !== undefined;
  }

  private isIUploadMessageVideoParameters(binaryOrParameters: number[] | IUploadMessageVideoParameters): binaryOrParameters is IUploadMessageVideoParameters {
    return (binaryOrParameters as IUploadMessageVideoParameters).binary !== undefined;
  }

  private isIAddMediaMetadataParameters(metadataOrParameters: IMediaMetadata | IAddMediaMetadataParameters): metadataOrParameters is IAddMediaMetadataParameters {
    return (metadataOrParameters as IAddMediaMetadataParameters).altText !== undefined;
  }
}
